### PR TITLE
[REVIEW] Add Fused L2 Expanded KNN kernel

### PR DESCRIPTION
**Pull request by mdoijade**
-- adds fused L2 expanded kNN kernel, this is faster by at least 20-25% on higher dimensions (D >= 128) than L2 unexpanded version.
-- also on smaller dimension (D <=32) L2 expanded is always faster by 10-15%
 -- slight improvement in updateSortedWarpQ device function by reducing redundant instruction.
-- Fix incorrect output for NN >32 case when taking prod-cons knn merge path, this was caught in HDBSCAN pytest.

### CODE REVIEW

File: ci/prtest.config
Comment (ajschmidt8) [inline ci/prtest.config:6]: The `ops` team is tagged as a `codeowner` for this file, but I'm not sure what the preferred value is for these variables. I'll defer to @dantegd for this change since it seems he's touched this file last.

**Comment by mdoijade** [comment]
@cjnolet @teju85 @dantegd  I think this PR is now in review state, please help with review whenever possible.

File: ci/prtest.config
Comment (cjnolet) [reply ci/prtest.config:6]: I don't believe this file is even being used currently but we should probably revert these changes so we're not introducing unneeded changes into this PR

File: cpp/include/raft/spatial/knn/detail/fused_l2_knn.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/detail/fused_l2_knn.cuh:331]: Aside from the template instantiations for each distance, is it possible that the unrolling of these loops could also be having a drastic increase on the compile time?

File: cpp/include/raft/spatial/knn/detail/knn_brute_force_faiss.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/detail/knn_brute_force_faiss.cuh:275]: The conditionals here are getting quite involved. It would be easier to read and maintain this if we moved everything inside this initial `if` block into a separate function and invoked that directly.

File: cpp/include/raft/spatial/knn/knn.hpp
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/knn.hpp:119]: This here is a potentially breaking change to libraries downstream. I know the expanded form can have quite a signifciant speedup from the exhausted unexpanded form but we'll want to at least verify this doesn't break cuml or euler before changing this. Another option could be to get this PR merged in sooner without this change and then make the change in a different PR once the downstream libraries have been tested.

File: cpp/test/spatial/ball_cover.cu
Comment (cjnolet) [inline cpp/test/spatial/ball_cover.cu:83]: Thanks for reverting this! Though we no longer need to invoke the fused knn directly, I do still benefit to keeping the additional helper function so it's more simple to change the bfknn call across all the gtests in the future.

File: cpp/test/spatial/ball_cover.cu
Comment (cjnolet) [inline cpp/test/spatial/ball_cover.cu:207]: Now that we have a knn that's not just proxying down to faiss, we should be gtesting it accordingly, similar to what's being done w/ the haversine and ball cover gtests. It's also important going forward because RAFT is beginning to get used by more projects and thus the impact of breaking tests is more than just cuml.

My suggestion is to test `l2_unexpanded_knn` and `l2_expanded_knn` directly in the gtests and then we can test the `brute_force_knn` more generally.

File: cpp/include/raft/spatial/knn/knn.hpp
Comment (mdoijade) [reply cpp/include/raft/spatial/knn/knn.hpp:119]: I'll be verifying this fork with cuML PR for verifying these changes in CI

File: cpp/test/spatial/ball_cover.cu
Comment (mdoijade) [reply cpp/test/spatial/ball_cover.cu:207]: certainly a rigorous tests within RAFT are needed for them, I'll add gtests for them separately, now instead of `l2_unexpanded_knn` and `l2_expanded_knn` we have single entry `fusedL2Knn` function for both of them.
For these kernel I relied on cuML cpp knn tests & pytests so far which is no longer correct as you rightly mention it.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by cjnolet** [comment]
add to allowlist

File: cpp/test/spatial/ball_cover.cu
Comment (mdoijade) [reply cpp/test/spatial/ball_cover.cu:207]: a new test `tests\spatial\fused_l2_knn.cu` is added which does testing for both L2 exp/unexp cases which compares its output with faiss bfknn call. 
I've also polished the fp32 atomicMax device function which I believe is more faster than atomicCAS based version and also takes care of NaNs.

Comment (mdoijade) [reply cpp/test/spatial/ball_cover.cu:207]: Apologies for the delay in updating this PR with the unit test

**Review by ChuckHastings — APPROVED** [review]
Shouldn't affect cugraph

**Comment by mdoijade** [comment]
@cjnolet can this PR get merged?

**Comment by mdoijade** [comment]
@cjnolet I see this PR is marked for v22.02 so you'll be auto merging it or there is additional action required from my side?

**Comment by cjnolet** [comment]
@mdoijade, I scraped through the PRs a couple weeks ago and aligned them to expected releases. Looking back through my reivew, I'm really happy for the new tests but it looks like there are still a couple (very minor) things to address.

**Comment by mdoijade** [comment]
> 
> 
> @mdoijade, I scraped through the PRs a couple weeks ago and aligned them to expected releases. Looking back through my reivew, I'm really happy for the new tests but it looks like there are still a couple (very minor) things to address.

@cjnolet  I believe now I have addressed all the points in this PR, the build failure is coming from `../test/sparse/dist_coo_spmv.cu`  in the CI, but locally I cannot repro this build failure.

**Review by cjnolet — APPROVED** [review]
LGTM

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**
